lib/github/api.nit

```diff
@@ -1602,13 +956,301 @@ end
 #
 # Mostly a wrapper around a json object.
 class GithubFile
-
-	# Json content.
-	var json: JsonObject
+	serialize
 
 	# File name.
-	fun filename: String do return json["filename"].as(String)
+	var filename: String
+end
 
-	# Set filename.
-	fun filename=(filename: String) do json["filename"] = filename
+# Make ISO Datew serilizable
+redef class ISODate
+	super Jsonable
+	serialize
+
+	redef fun to_json do return serialize_to_json
 end
+
+# JsonDeserializer specific for Github objects.
+class GithubDeserializer
+	super JsonDeserializer
+
+	redef fun class_name_heuristic(json_object) do
+		if json_object.has_key("login") or json_object.has_key("email") then
+			return "User"
+		else if json_object.has_key("full_name") then
+			return "Repo"
+		else if json_object.has_key("name") and json_object.has_key("commit") then
+			return "Branch"
+		else if json_object.has_key("sha") and json_object.has_key("ref") then
+			return "PullRef"
+		else if json_object.has_key("sha") or (json_object.has_key("id") and json_object.has_key("tree_id")) then
+			return "Commit"
+		else if json_object.has_key("number") and json_object.has_key("patch_url") then
+			return "PullRequest"
+		else if json_object.has_key("open_issues") and json_object.has_key("closed_issues") then
+			return "Milestone"
+		else if json_object.has_key("number") and json_object.has_key("title") then
+			return "Issue"
+		else if json_object.has_key("color") then
+			return "Label"
+		else if json_object.has_key("event") then
+			return "IssueEvent"
+		else if json_object.has_key("original_commit_id") then
+			return "ReviewComment"
+		else if json_object.has_key("commit_id") then
+			return "CommitComment"
+		else if json_object.has_key("issue_url") then
+			return "IssueComment"
+		end
+		return null
+	end
+
+    redef fun deserialize_class(name) do
+        if name == "Repo" then return deserialize_repo
+		if name == "User" then return deserialize_user
+		if name == "Branch" then return deserialize_branch
+		if name == "Commit" then return deserialize_commit
+		if name == "Issue" then return deserialize_issue
+		if name == "PullRequest" then return deserialize_pull_request
+		if name == "Label" then return deserialize_label
+		if name == "Milestone" then return deserialize_milestone
+		if name == "PullRef" then return deserialize_pullref
+		if name == "IssueEvent" then return deserialize_issueevent
+		if name == "CommitComment" then return deserialize_commit_comment
+		if name == "ReviewComment" then return deserialize_review_comment
+		if name == "IssueComment" then return deserialize_issue_comment
+		return super
+	end
```

Comment:
You would need less code by leaving the generated code do its work. Notably the method `deserialize_class` could be shorter and you wouldn't need the other `deserialize_*` methods.

The method `deserialize_class` generated at compilation looks very much like yours, but it calls the deserialization constructor:

~~~
# Approximately what is generated
redef fun deserialize_class(name) do
	if name == "Repo" then return new Repo.from_deserializer(self)
	if name == "User" then return new User.from_deserializer(self)
...
~~~

And the generated deserialization constructors `from_deserializer` also have pretty much the same behavior as the method `deserialize_repo` and others `deserialize_*`.

In theory, removing the `deserialize_class` method and the other `deserialize_*` will fallback on the generated code and it would have the same behavior. (with a few bonuses like error handling)

However, the attributes typed with `ISODate` are a problem. Your custom methods `deserialize_*` avoid this issue but the other classes (without a `ISODate`) could be left to the generated services to get the desired behavior.

An alternative to handle `ISODate` would be to keep the dates as strings and use a getter, you could then remove all the custom `deserialize_*`!
~~~
class Commit
    serialize

    private var author_date_string: String is serialize_as("author_date")

    fun author_date: ISODate do return new ISODate.from_string(author_date_string)
    # or: var author_date = new ISODate.from_string(author_date_string) is lazy, noserialize
~~~